src/topology/bounded_continuous_function.lean

```diff
@@ -429,5 +429,112 @@ def of_normed_group_discrete {α : Type u} {β : Type v}
+begin
+  rw [← sub_zero f, ← dist_eq_norm, bounded_continuous_function.dist_eq],
+  simp,
+end
+
+
 end normed_group
+
+
+section normed_space
+/-! In this section, if `β` is a normed space, then we show that the space of bounded
+continuous functions from `α` to `β` inherits a normed space structure, by using
+pointwise operations and checking that they are compatible with the uniform distance. -/
+
+variables [nondiscrete_normed_field 𝕜]
+variables [topological_space α] [normed_group β] [normed_space 𝕜 β]
+variables {f g : α →ᵇ β} {x : α} {C : ℝ}
+
+instance : has_scalar 𝕜 (α →ᵇ β) :=
+⟨λc, λf, ⟨λx, c • f.1 x,
+  continuous.smul continuous_const f.2.left,
+  begin
+    cases f.2.right with C hbound,
+    use ∥c∥ * C,
+    intros,
+    have hnneg : 0 ≤ ∥c∥,
+    { exact norm_nonneg c },
+    specialize hbound x y,
+    rw dist_eq_norm at hbound ⊢,
+      calc ∥c • f x - c • f y∥ = ∥c • (f x - f y)∥ : by rw smul_sub c (f x) (f y)
+    ... = ∥c∥ * ∥f x - f y∥ : norm_smul c (f x - f y)
+    ... ≤ ∥c∥ * C : mul_le_mul_of_nonneg_left hbound hnneg,
+  end⟩⟩
+
+
+instance : module 𝕜 (α →ᵇ β) :=
+  module.of_core $
+  { smul := (•),
+    smul_add := begin
+    intros c f g, ext,
+    exact smul_add c (f x) (g x)
+    end,
+    add_smul := begin
+    intros c₁ c₂ f, ext,
+    exact add_smul c₁ c₂ (f x),
+    end,
+    mul_smul := begin
+    intros c₁ c₂ f,
+    ext, exact mul_smul c₁ c₂ (f x),
+    end,
+    one_smul := begin
+    intros f,
+    ext, exact one_smul 𝕜 (f x),
+    end }
+
+
+
+instance : vector_space 𝕜 (α →ᵇ β) :=
+{ .. bounded_continuous_function.module }
+
+
+lemma bounded_continuous_sub_smul (c : 𝕜) (f : α →ᵇ β) :  ∥c • f∥ ≤ ∥c∥ * ∥f∥ :=
+begin
+  have hnneg : 0 ≤ ∥ c ∥,
+    exact norm_nonneg c,
+  rw norm_eq (c • f),
+  apply real.Inf_le,
+  { use 0, intros y hy,
+    rw set.mem_set_of_eq at hy,
+    exact hy.left },
+  { rw set.mem_set_of_eq,
+    split,
+    { exact mul_nonneg' hnneg (norm_nonneg f) },
+    { intros,
+      calc ∥(c • f) x∥ = ∥c∥ * ∥f x∥ : norm_smul c (f x)
+      ... ≤ ∥c∥ * ∥f∥ : mul_le_mul_of_nonneg_left (norm_coe_le_norm x) hnneg } }
+end
+
+lemma bounded_continuous_smul (c : 𝕜) (f : α →ᵇ β) : ∥c • f∥ = ∥c∥ * ∥f∥ :=
+begin
+  by_cases h : c = 0,
+  rw h, simp,
+  have hnneg : 0 ≤ ∥c∥,
+    exact norm_nonneg c,
+  apply le_antisymm,
+  exact bounded_continuous_sub_smul c f,
+  have hinv : ∥f∥ ≤ ∥1 / c∥ * ∥c • f∥,
+  { calc ∥f ∥= ∥(1 : 𝕜) • f∥ : by simp
+    ... = ∥(1 / c * c ) • f∥ : by rw (div_mul_cancel 1 h)
+    ... = ∥(1 / c) • ( c • f)∥ : by rw (mul_smul _ _ _).symm
+    ... ≤ ∥1 / c∥ * ∥c • f∥ : bounded_continuous_sub_smul (1 / c) (c • f) },
+  calc ∥c∥ * ∥f∥  ≤ ∥c∥ * (∥1 / c∥ * ∥c • f∥) : mul_le_mul_of_nonneg_left hinv hnneg
+  ... = (∥c ∥ * ∥1 / c∥) * ∥c • f∥ : by ring
+  ... = ∥c * (1 / c)∥ * ∥c • f∥ : by rw (normed_field.norm_mul c (1/c))
+  ... = ∥(1 : 𝕜)∥ * ∥c • f∥ : by rw (mul_div_cancel' 1 h)
+  ... = 1 * ∥c • f∥ : by rw normed_field.norm_one
+  ... = ∥c • f∥ : by ring,
+end
+
```

Comment:
```suggestion
  { rw [h, zero_smul, norm_zero, norm_zero, zero_mul], },
  { have hnneg : 0 ≤ ∥c∥ := norm_nonneg c,
    apply le_antisymm,
    { exact bounded_continuous_sub_smul c f, },
    { have hinv : ∥f∥ ≤ ∥1 / c∥ * ∥c • f∥,
      { calc ∥f ∥= ∥(1 : 𝕜) • f∥ : by rw one_smul
        ... = ∥(1 / c * c ) • f∥ : by rw (div_mul_cancel 1 h)
        ... = ∥(1 / c) • ( c • f)∥ : by rw (mul_smul _ _ _).symm
        ... ≤ ∥1 / c∥ * ∥c • f∥ : bounded_continuous_sub_smul (1 / c) (c • f) },
      calc ∥c∥ * ∥f∥  ≤ ∥c∥ * (∥1 / c∥ * ∥c • f∥) : mul_le_mul_of_nonneg_left hinv hnneg
      ... = (∥c ∥ * ∥1 / c∥) * ∥c • f∥ : by rw mul_assoc
      ... = ∥c * (1 / c)∥ * ∥c • f∥ : by rw (normed_field.norm_mul c (1/c))
      ... = ∥(1 : 𝕜)∥ * ∥c • f∥ : by rw (mul_div_cancel' 1 h)
      ... = 1 * ∥c • f∥ : by rw normed_field.norm_one
      ... = ∥c • f∥ : by rw one_mul, } }
end
```
When new goals appear in a proof, we usually surround the statements that prove each of the goals with "focus braces".

I also replaced some `simp` and `ring` invocations with explicit rewrites to speed up the proof a bit (from 2.78s to 663ms according to `set_option profiler true`). Stylistically, I think usually `simp` and `ring` are preferred even if they makes things a little bit slower, since they can hide a lot of fiddly details. In this case, I think the speedup is worth a bit of verbosity.

Reply:
I had been wondering how to find this balance/make this call.  Thank you for explaining your reasoning.